### PR TITLE
lint: add checker for unexported method calls on unexported types

### LINKED ISSUE #92: lint: add checker for unexported method calls on unexported types
Calling unexported method on unexported type is not an error, but it's nicer
to make methods that are intended to be called from outside of type itself exported (`golint` doesn't require having doc-comments for them, so it doesn't add a burden).

This makes refactoring a type into exported later easier.

This code triggers a warning:
```go
package main

type foo struct{}

func (foo) unexported() {}
func (foo) Exported() {}

func main() {
  var f foo
  f.unexported() // Warning
  f.Exported() // OK, call of exported method
}
```

### PR DESCRIPTION
**Pull request by fexolm**
fixes #92

### CODE REVIEW

File: cmd/kfulint/testdata/unexported-call/checker_tests.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+func getFoo() foo { return foo{} }
+
+func f() {
+	var fo foo
+	///: getFoo().unexported() should be exported
+	getFoo().unexported()
+	///: fo.unexported() should be exported
+	fo.unexported()
+	fo.Exported()
+}
```
Comment (Quasilyte) [inline cmd/kfulint/testdata/unexported-call/checker_tests.go]: Need a test case for `foo` method that calls unexported methods and gets no warnings.

File: lint/unexported-call_checker.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	for _, decl := range collectFuncDecls(f) {
+		if decl.Body == nil {
+			continue
+		}
+		name := ""
+		cond := decl.Recv != nil &&
+			decl.Recv.List != nil &&
+			len(decl.Recv.List) == 1 &&
+			decl.Recv.List[0].Names != nil &&
+			len(decl.Recv.List[0].Names) == 1
```
Comment (Quasilyte) [inline lint/unexported-call_checker.go]: I believe two nil checks for `List` fields are redundant as nil slice length is 0.
The `== 1` check would cover both cases and is perfectly safe.

Comment (Quasilyte) [reply lint/unexported-call_checker.go]: The `gosimple` doesn't do a best job with it's [S1009](https://staticcheck.io/docs/gosimple#S1009).
It should catch these cases as well.

File: lint/unexported-call_checker.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	typ := c.ctx.TypesInfo.TypeOf(sel.Sel)
+	sig, ok := typ.(*types.Signature)
+	if !ok {
+		return
+	}
+
+	if sig.Recv() == nil || sig.Recv().Type() == nil {
+		return
+	}
+	typ2, ok := sig.Recv().Type().(*types.Named)
```
Comment (Quasilyte) [inline lint/unexported-call_checker.go]: Maybe `s/typ2/recvTyp/`?

**Comment by Quasilyte** [comment]
You've said that it handles GOROOT without crashes?
If that's true, LGTM when comments are addressed.

**Comment by fexolm** [comment]
@will merge it after #109

File: lint/unexported-call_checker.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import (
+	"fmt"
+	"go/ast"
+	"go/types"
+)
+
+type unexportedCallChecker struct {
+	baseLocalExprChecker
+
+	funcName string
```
Comment (Quasilyte) [inline lint/unexported-call_checker.go]: Should be `recvName`.

File: lint/unexported-call_checker.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	if cond {
+		c.funcName = decl.Recv.List[0].Names[0].Name
+	}
+	return true
+}
+
+// Check finds calls of unexported method from unexported type
+// outside that type.
+//
+//TODO: update description and warning message
```
Comment (Quasilyte) [inline lint/unexported-call_checker.go]: Please put a spece after `//`.
Comments without a space are treated as "pragmas" in Go gc toolchain.
Especially if "keyword" ends with `:`.

**Comment by fexolm** [comment]
@Quasilyte, fixed.

**Pull request merged**